grants_tagger/train_with_sagemaker.py

```diff
@@ -0,0 +1,132 @@
+    code_path: path to addtitional source files needed for train
+    model_path: path to save model
+    data_path: path to data
+    label_binarizer_path: path to save label_binarizer
+    params: additional params that are passed to train
+
+    Note that paths need to be converted to container paths. Sagemaker is responsible for moving
+    data and models in an out of the container.
+    Also params are equivalent to the rest of the parameters that are passed through the CLI to 
+    train. Sagemaker stores all those into hyperapameters and converts to CLI args upon invokation
+
+    Returns:
+    hyperaparameters
+    """
+    # Step 1 - Convert data, model paths
+    _, model_subpath = os.path.split(model_path) if model_path else ("","")# can be dir or model_name
+    _, data_filename = os.path.split(data_path)
+    _, label_binarizer_filename = os.path.split(label_binarizer_path)
+
+    container_model_path = f"/opt/ml/model/{model_subpath}" if model_subpath else ""
+    container_data_path = f"/opt/ml/input/data/training/{data_filename}"
+    container_label_binarizer_path = f"/opt/ml/model/{label_binarizer_filename}"
+
+    hyperparameters = {
+        "data_path": container_data_path,
+        "model_path": container_model_path,
+        "label_binarizer_path": container_label_binarizer_path
+    }
+
+    # Step 2 - Add additional params passed to train in hyperparams
+    for k, v in params.items():
+        hyperparameters[k] = v
+    
+    # Step 3 - Define entrypoint to invoke and additional source files needed
+    hyperparameters["sagemaker_program"] = entrypoint
+    hyperparameters["sagemaker_submit_directory"] = code_path
+    
+    # Step 4 - Package hyperparams params in json
+    hyperparameters = {str(k): json.dumps(v) for (k, v) in hyperparameters.items() if v}
+    return hyperparameters
+    
+def train_with_sagemaker(instance_type="local", config_version=None, **kwargs):
+    """Invokes train with kwargs using Sagemaker using instance type"""
+    code_path = upload_code(ENTRYPOINT, DEPENDENCIES, bucket=PROJECTS_BUCKET, prefix=PROJECT_NAME)
+    logger.info(code_path)
+
+    hyperparameters = create_hyperparameters(ENTRYPOINT, code_path, **kwargs)
+    logger.info(hyperparameters)
+
+    data_path = kwargs["data_path"]
+    if instance_type == "local":
+        model_path = f"file://"
+        data_path = f"file://{data_path}"
+    else:
+        model_path = f"s3://{PROJECTS_BUCKET}/{PROJECT_NAME}/output/"
+        data_path = f"s3://{PROJECTS_BUCKET}/{PROJECT_NAME}/{data_path}"
+    logger.info(f"model_path: {model_path}")
+    logger.info(f"data_path: {data_path}")
+
+    base_job_name = kwargs["approach"]
+    if config_version:
+        config_version = config_version.replace(".","-")
+        base_job_name = f"{base_job_name}-{config_version}"
+    es = Estimator(
+        image_uri=os.environ["ECR_IMAGE"],
+        role=SAGEMAKER_ROLE,
+        framework_version="0.20.0",
+        instance_count=1,
```

Comment:
There is nothing here in the class `Estimator` that tells the codepath. Does SageMaker have a specific path it looks at?

Reply:
It looks at `/opt/ml/code` inside the container and copies the code from the s3 path defined in the hyperparameters `hyperparameters["sagemaker_submit_directory"] = code_path` there